src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -39,79 +40,76 @@ import scala.annotation.tailrec
-    override def run(): Unit = {
-      require(_tasksLocal.get eq null)
+    def this(r: Runnable) = {
+      this(4)
+      addLast(r)
+    }
 
-      val prevBlockContext = BlockContext.current
-      BlockContext.withBlockContext(this) {
-        try {
-          parentBlockContext = prevBlockContext
+    final def executor: BatchingExecutor = BatchingExecutor.this
 
-          @tailrec def processBatch(batch: List[Runnable]): Unit = batch match {
-            case Nil => ()
-            case head :: tail =>
-              _tasksLocal set tail
-              try {
-                head.run()
-              } catch {
-                case t: Throwable =>
-                  // if one task throws, move the
-                  // remaining tasks to another thread
-                  // so we can throw the exception
-                  // up to the invoking executor
-                  val remaining = _tasksLocal.get
-                  _tasksLocal set Nil
-                  unbatchedExecute(new Batch(remaining)) //TODO what if this submission fails?
-                  throw t // rethrow
-              }
-              processBatch(_tasksLocal.get) // since head.run() can add entries, always do _tasksLocal.get here
-          }
+    // this method runs in the delegate ExecutionContext's thread
+    override final def run(): Unit = BlockContext.usingBlockContext(this)(this)
+
+    override final def apply(prevBlockContext: BlockContext): Unit = {
+      //This invariant needs to hold: require(_tasksLocal.get eq null)
+      parentBlockContext = prevBlockContext
+      try {
+        _tasksLocal.set(this)
+        runAll()
+        _tasksLocal.remove() // Will be cleared in the throwing-case by runAll()
+      } finally {
+        parentBlockContext = null
+      }
+    }
 
-          processBatch(initial)
-        } finally {
-          _tasksLocal.remove()
-          parentBlockContext = null
-        }
+    @tailrec private[this] final def runAll(): Unit = {
+      val next = pollLast()
+      if (next ne null) {
+        try next.run() catch {
+          case t: Throwable =>
+            parentBlockContext = null // Need to reset this before re-submitting it
+            _tasksLocal.remove() // If unbatchedExecute runs synchronously
+            unbatchedExecute(this) //TODO what if this submission fails?
+            throw t
+         }
+        runAll()
       }
     }
 
     override def blockOn[T](thunk: => T)(implicit permission: CanAwait): T = {
-      // if we know there will be blocking, we don't want to keep tasks queued up because it could deadlock.
-      {
-        val tasks = _tasksLocal.get
-        _tasksLocal set Nil
-        if ((tasks ne null) && tasks.nonEmpty)
-          unbatchedExecute(new Batch(tasks))
+      val pbc = parentBlockContext
+      if(!isEmpty) { // if we know there will be blocking, we don't want to keep tasks queued up because it could deadlock.
+        val b = new Batch(math.max(4, this.size))
```

Comment:
I tested it in isolation so I'm not sure, but it seems that the Scala compiler won't resolve to the `size` method in `ArrayDeque` and create a field in `Batch`, which is unnecessary. Maybe rename the constructor param to `_size`. 

```scala
class Batch(size: Int) extends ArrayDeque(size) {
  def test = this.size
}
```
Decompiled:
```java
@ScalaSignature(bytes="\u0006\u0001\t2A!\u0001\u0002\u0001\u000b\t)!)\u0019;dQ*\t1!\u0001\u0006qY\u0006LxM]8v]\u0012\u001c\u0001a\u0005\u0002\u0001\rA\u0019q\u0001\u0004\b\u000e\u0003!Q!!\u0003\u0006\u0002\tU$\u0018\u000e\u001c\u0006\u0002\u0017\u0005!!.\u0019<b\u0013\ti\u0001B\u0001\u0006BeJ\f\u0017\u0010R3rk\u0016\u0004\"a\u0004\n\u000e\u0003AQ\u0011!E\u0001\u0006g\u000e\fG.Y\u0005\u0003'A\u0011qAT8uQ&tw\r\u0003\u0005\u0016\u0001\t\u0005\t\u0015!\u0003\u0017\u0003\u0011\u0019\u0018N_3\u0011\u0005=9\u0012B\u0001\r\u0011\u0005\rIe\u000e\u001e\u0005\u00065\u0001!\taG\u0001\u0007y%t\u0017\u000e\u001e \u0015\u0005qq\u0002CA\u000f\u0001\u001b\u0005\u0011\u0001\"B\u000b\u001a\u0001\u00041\u0002\"\u0002\u0011\u0001\t\u0003\t\u0013\u0001\u0002;fgR,\u0012A\u0006")
public class Batch
extends ArrayDeque<Nothing$> {
    private final int size;

    public int test() {
        return this.size;
    }

    public Batch(int size) {
        this.size = size;
        super(size);
    }
}

```

Reply:
@fwbrasil I renamed the parameter to `capacity`. It worked to do `super[ArrayDeque].size` but …